### PR TITLE
HIP-584: Copy approve allowance checks from services

### LINKED ISSUE #5652: HIP-584: Copy implementation for approve precompiles
### Problem

We need default implementations for precompiles related to approve logic, so we should copy needed logic from hedera-services into hedera-mirror-web3.

### Solution

Copy precompile logic for the following precompiles:

- ApprovePrecompile

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by 0xivanov**
**Description**:
For Approve precompiles we need to copy approve allowance checks from services.

**Related issue(s)**:

Fixes #5652

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/ApproveAllowanceChecks.java [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+     * @return response code after validation
+     */
+    public ResponseCodeEnum allowancesValidation(
+            final List<CryptoAllowance> cryptoAllowances,
+            final List<TokenAllowance> tokenAllowances,
+            final List<NftAllowance> nftAllowances,
+            final Account payerAccount,
+            final Store store) {
+        // feature flag for allowances
+        if (!isEnabled()) {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/ApproveAllowanceChecks.java:63]: I think we won't need this feature flag check and will always support allowances in the context of Archive Node

File: hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/DeleteAllowanceChecks.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+     *
+     * @param nftAllowances given nft serials allowances to remove
+     * @param payerAccount payer for the transaction
+     * @param store store
+     * @return validation response
+     */
+    public ResponseCodeEnum deleteAllowancesValidation(
+            final List<NftRemoveAllowance> nftAllowances, final Account payerAccount, final Store store) {
+        // feature flag for allowances
+        if (!isEnabled()) {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/DeleteAllowanceChecks.java:53]: Again, we can delete this feature flag check

File: hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/ApproveAllowanceChecks.java [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+import com.hedera.services.store.models.Token;
+import com.hedera.services.utils.EntityNum;
+import com.hederahashgraph.api.proto.java.CryptoAllowance;
+import com.hederahashgraph.api.proto.java.NftAllowance;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.hederahashgraph.api.proto.java.TokenAllowance;
+import com.hederahashgraph.api.proto.java.TokenSupplyType;
+import java.util.List;
+
+public class ApproveAllowanceChecks extends AllowanceChecks {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/ApproveAllowanceChecks.java:44]: You can document in javadoc the differences from hedera-services, e.g using the Store interface and passing it as an argument.

File: hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/AllowanceChecks.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import java.util.HashSet;
+import java.util.List;
+import org.apache.commons.lang3.tuple.Pair;
+
+/**
+ * Validations for {@link com.hederahashgraph.api.proto.java.CryptoApproveAllowance} transaction
+ * allowances
+ */
+public class AllowanceChecks {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/AllowanceChecks.java:43]: You can document in javadoc the differences from hedera-services, e.g using the Store interface and passing it as an argument.

File: hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/ApproveAllowanceChecks.java (hunk 1 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/ApproveAllowanceChecks.java]: Fixed

File: hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/AllowanceChecks.java (hunk 4 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/AllowanceChecks.java:46]: Fixed

File: hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/ApproveAllowanceChecks.java (hunk 3 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/ApproveAllowanceChecks.java:51]: Fixed

File: hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/DeleteAllowanceChecks.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import com.hedera.services.store.models.Id;
+import com.hedera.services.store.models.Token;
+import com.hederahashgraph.api.proto.java.NftRemoveAllowance;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import java.util.List;
+
+/**
+ * Semantic check validation for {@link com.hederahashgraph.api.proto.java.CryptoDeleteAllowance}
+ * transaction
+ */
```
[the remark this answers is not in the document]
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/services/txns/crypto/validators/DeleteAllowanceChecks.java:41]: Fixed

**Review by steven-sheehy — APPROVED** [review]
LGTM. Can merge once Ivan approves

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM

**Pull request merged**